torch/optim/adagrad.py

```diff
@@ -321,7 +321,11 @@ def _multi_tensor_adagrad(
     grouped_tensorlists = Optimizer._group_tensors_by_device_and_dtype([params, grads, state_sums, state_steps])
     for ((device_params, device_grads, device_state_sums, device_state_steps), _) in grouped_tensorlists.values():
 
-        device_has_sparse_grad = any(grad.is_sparse for grad in device_grads)
```

Comment:
Don't think this actually helps. In all likelihood, using global `has_sparse_grad` is sufficient.

Reply:
The semantic difference here is that if after grouping there are groups with sparse tensors and other groups that don't have sparse tensors, this code in eager would allow more foreach performance to happen (vs falling back to for loop every time). 

I do suspect that for most cases, all params are either sparse or not sparse, so the global could work heuristically.

Something I had thought about before that would be interesting is to have the grouping logic group sparse tensors (and all mta invalid tensors) into device cpu. Then, we could do away with this variable entirely. Now that grouping's been done in C++, it would not be as trivial to add that, but that seems worth exploring too.